moveit_planners/pilz_industrial_motion_planner/src/plan_components_builder.cpp

```diff
@@ -62,7 +67,14 @@ void PlanComponentsBuilder::appendWithStrictTimeIncrease(robot_trajectory::Robot
   if (!pilz_industrial_motion_planner::isRobotStateEqual(result.getLastWayPoint(), source.getFirstWayPoint(),
                                                          result.getGroupName(), ROBOT_STATE_EQUALITY_EPSILON))
   {
-    result.append(source, source.getWayPointDurationFromStart(0));
+    if (source.getWayPointDurationFromPrevious(0) == 0)
+    {
+      result.append(source, source.getWayPointDurationFromStart(0));
+    }
+    else
```

Comment:
Can this be instead added to the condition in line 62?

```cpp
if (result.empty() || source.getWayPointDurationFromPrevious(0) > 0.0) {
    result.append(source, 0.0);
    return;
}
```

It may make this logic a bit easier to follow

Reply:
Alternatively, do things still work out if instead of returning early in the IF case here, you simply let the function return at the very end?

Reply:
Hmm, you're right. Maybe we shouldn't do that.

Let me test with this and get back to you.

Reply:
OK I tested on [this tutorial example](https://moveit.picknik.ai/main/doc/how_to_guides/pilz_industrial_motion_planner/pilz_industrial_motion_planner.html#sequence-of-multiple-segments) and unfortunately got this error:

```
[ros2_control_node-5] [INFO] [1734435722.902666137] [panda_arm_controller]: Received new action goal
[ros2_control_node-5] [ERROR] [1734435722.902766306] [panda_arm_controller]: Time between points 27 and 28 is not strictly increasing, it is 2.700000 and 2.700000 respectively
[move_group-4] [INFO] [1734435722.903129429] [moveit.simple_controller_manager.follow_joint_trajectory_controller_handle]: panda_arm_controller started execution
[move_group-4] [WARN] [1734435722.903173172] [moveit.simple_controller_manager.follow_joint_trajectory_controller_handle]: Goal request rejected
```

In this example, both duration from start and previous are `0.1`, so the command `result.append(source, 0.0);` will not work out.

What are the duration from start and previous in your case? Wonder if there is some other special case happening on your end?

Reply:
The thing is, that on Humble, `ros2_control`'s joint trajectory controller actually allowed duplicate waypoints to be at the same time and it would deconflict them. In later versions, errors like the above show up.

So appending waypoints whose state is not equal, with zero time offset, is not something we should be doing.

Reply:
I actually don't know what changed between Humble and later versions that having these duplicate time points ends up working... since these lines of code have been there since Humble:

https://github.com/ros-controls/ros2_controllers/blob/0438a1e886386035295da90c3e3dc0436217ddcc/joint_trajectory_controller/src/joint_trajectory_controller.cpp#L1444-L1452

Reply:
Oh INTERESTING! It seems that the Humble branch does not have this commit: https://github.com/moveit/moveit2/pull/1813

which may be why it works there, but not on later versions.

Reply:
Yeah, after I include #1821 my results look good!
I think I'll  just close my PR and waiting for #1821 to be merged to humble.
Thanks for your help.

Reply:
Thanks for confirming! Will do this now.